### PR TITLE
Arrays

### PR DESCRIPTION
**Pull request by HolyHol**

### CODE REVIEW

File: test/exercise/arrays/solution.rb [hunk 1]
```diff
@@ -1,12 +1,69 @@
...
+      def max(array)
+        return nil if array.empty?
+        
+        if array.length == 1
+          array[0] # base case
+        else
+          sub_max = max(array[1..-1])
+          array[0] > sub_max ? array[0] : sub_max
+        end
+      end
```
Comment (nut4k1) [inline test/exercise/arrays/solution.rb]: First, this methods are identical
Second, in Ruby if there are two methods with the same name, last one will be used (even if they have different interfaces. e.g. in Java it's called _overload_, I suppose).

File: test/exercise/arrays/solution.rb [hunk 2]
```diff
@@ -1,12 +1,69 @@
...
+      
+      def replaces_positive_array(array, max_array_new)
+
```
Comment (nut4k1) [inline test/exercise/arrays/solution.rb]: no need empty line before **return**, check with rubocop

File: test/exercise/arrays/solution.rb [hunk 3]
```diff
@@ -1,12 +1,69 @@
 module Exercise
   module Arrays
     class << self
+      def max(list)
+        return nil if list.empty?
+      
+        if list.length == 1
+          list[0] # base case
```
Comment (nut4k1) [inline test/exercise/arrays/solution.rb]: no need to check for that case, algorithm suppose to work with all valid cases

File: test/exercise/arrays/solution.rb [hunk 4]
```diff
@@ -1,12 +1,69 @@
...
+          sub_max = max(array[1..-1])
+          array[0] > sub_max ? array[0] : sub_max
+        end
+      end
+      
+      def replaces_positive_array(array, max_array_new)
+        
+        return [] if array.empty?
+        
+        sub_array = replaces_positive_array(array[1..-1], max_array_new)
```
Comment (nut4k1) [inline test/exercise/arrays/solution.rb]: you are loosing first element here

File: test/exercise/arrays/solution.rb [hunk 5]
```diff
@@ -1,12 +1,69 @@
...
+        
+        if array[0] < 0
+          sub_array.unshift(array[0])
+          
+        else 
+          sub_array.unshift(max_array_new) 
+          
+        end
```
Comment (nut4k1) [inline test/exercise/arrays/solution.rb]: please, run rubocop for style checks & fixes

File: test/exercise/arrays/solution.rb [hunk 6]
```diff
@@ -1,12 +1,69 @@
...
+            return mid 
+          
+          elsif guess > _query
+            high = mid - 1
+          
+          else 
+            low = mid + 1
+          end
+        end
+      return -1
```
Comment (nut4k1) [inline test/exercise/arrays/solution.rb]: main idea of this binary search:
-  data is sorted
- complexity (max iterations) should be equal to **log2(N)**

base on your loop condition `while low <= high` it is **N-1**

P.S.: you can check it by using count variable and increasing it with each iteration

File: test/exercise/arrays/solution.rb [hunk 7]
```diff
@@ -1,12 +1,69 @@
...
+
+      def max(array)
+        return nil if array.empty?
+        
+        if array.length == 1
+          array[0] # base case
+        else
+          sub_max = max(array[1..-1])
+          array[0] > sub_max ? array[0] : sub_max
+        end
```
Comment (nut4k1) [inline test/exercise/arrays/solution.rb]: you no need to use recursive here

**Review by nut4k1 — CHANGES_REQUESTED** [review]
Mainly looks good. Try to implement algorithm, not to create it.

**Comment by nut4k1** [comment]
hints if needed

https://gist.github.com/nut4k1/fb31e79e24998934e82b0ade3f973a1e

File: test/exercise/arrays/solution.rb (hunk 3 above)
Comment (HolyHol) [reply test/exercise/arrays/solution.rb]: Sorry, i didn't get

File: test/exercise/arrays/solution.rb (hunk 4 above)
Comment (HolyHol) [reply test/exercise/arrays/solution.rb]: In which cases? Tests are passing

File: test/exercise/arrays/solution.rb (hunk 3 above)
Comment (nut4k1) [reply test/exercise/arrays/solution.rb]: code should work with any size. check my attachment below
or text me in tg

File: test/exercise/arrays/solution.rb (hunk 4 above)
Comment (nut4k1) [reply test/exercise/arrays/solution.rb:30]: I've copied your code and it failed

Comment (nut4k1) [reply test/exercise/arrays/solution.rb]: yep, sorry. I've tried again and it works. but still using recursive here is not a best option

File: test/exercise/arrays/solution.rb (hunk 3 above)
Comment (HolyHol) [reply test/exercise/arrays/solution.rb]: It's working, I can't get the issue

Comment (nut4k1) [reply test/exercise/arrays/solution.rb]: It's working. But the way HOW it works also matters

File: test/exercise/arrays/solution.rb (hunk 6 above)
Comment (HolyHol) [reply test/exercise/arrays/solution.rb]: i check and it's working like log2(N)

```ruby
irb(main):001:1* def search(_array, _query)
irb(main):002:1*   # Реализовать двоичный поиск
irb(main):003:1*   # Функция должна возвращать индекс элемента
irb(main):004:1*   low = 0
irb(main):005:1*   high = _array.length - 1
irb(main):006:1*   count = 0
irb(main):007:1*
irb(main):008:2*   while low <= high
irb(main):009:2*     mid = (low + high) / 2
irb(main):010:2*     guess = _array[mid]
irb(main):011:2*     count = count + 1
irb(main):012:2*
irb(main):013:3*     if guess == _query
irb(main):014:3*       puts count
irb(main):015:3*       return mid
irb(main):016:3*
irb(main):017:3*     elsif guess > _query
irb(main):018:3*       high = mid - 1
irb(main):019:3*
irb(main):020:3*     else
irb(main):021:3*       low = mid + 1
irb(main):022:2*     end
irb(main):023:1*   end
irb(main):024:1*   -1
irb(main):025:0> end
=> :search
irb(main):026:0> search((1..10000).to_a, 1000)
11
=> 999
irb(main):027:0>
```

Comment (nut4k1) [reply test/exercise/arrays/solution.rb]: so just change a condition for the loop :)

Comment (HolyHol) [reply test/exercise/arrays/solution.rb]: I really don't understand what do u mean :(

Comment (nut4k1) [reply test/exercise/arrays/solution.rb]: ok, then just apply a few style fixes and roll forward

**Review by nut4k1 — APPROVED** [review]

**Pull request merged**